packages/fuels-programs/src/retry.rs

```diff
@@ -0,0 +1,455 @@
+        async fn returns_last_error() -> fuels_core::types::errors::Result<()> {
+            let err_msgs = ["Err1", "Err2", "Err3"];
+            let number_of_attempts = Mutex::new(0usize);
+
+            let will_always_fail = || async {
+                let msg = err_msgs[*number_of_attempts.lock().await];
+                *number_of_attempts.lock().await += 1;
+
+                Result::<(), _>::Err(Error::InvalidData(msg.to_string()))
+            };
+
+            let should_retry_fn = |_res: &_| -> bool { true };
+
+            let retry_options = RetryConfig::new(3, Backoff::Linear(Duration::from_millis(10)));
+
+            let err = retry(will_always_fail, &retry_options, should_retry_fn)
+                .await
+                .expect_err("Should have failed");
+
+            assert_eq!(
+                err.to_string(),
+                Error::InvalidData(err_msgs[2].to_string()).to_string()
+            );
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn returns_value_on_success() -> anyhow::Result<()> {
+            let values = Mutex::new(vec![
+                Ok(String::from("Success")),
+                Err(Error::InvalidData("Err1".to_string())),
+                Err(Error::InvalidData("Err2".to_string())),
+            ]);
+
+            let will_always_fail = || async { values.lock().await.pop().unwrap() };
+
+            let should_retry_fn = |res: &_| -> bool {
+                matches!(res, Err(err) if matches!(err, Error::InvalidData(_)))
+            };
+
+            let retry_options = RetryConfig::new(3, Backoff::Linear(Duration::from_millis(10)));
+
+            let ok = retry(will_always_fail, &retry_options, should_retry_fn).await?;
+
+            assert_eq!(ok, "Success");
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn retry_on_none_values() -> anyhow::Result<()> {
+            let values = Mutex::new(vec![
+                Ok::<Option<String>, Error>(Some(String::from("Success"))),
+                Ok(None),
+                Ok(None),
+            ]);
+            let will_always_fail = || async { values.lock().await.pop().unwrap() };
+
+            let should_retry_fn = |res: &_| -> bool {
+                match res {
+                    Err(err) if matches!(err, Error::IOError(_)) => true,
+                    Ok(None) => true,
+                    _ => false,
+                }
+            };
+
+            let retry_options = RetryConfig::new(3, Backoff::Linear(Duration::from_millis(10)));
+
+            let ok = retry(will_always_fail, &retry_options, should_retry_fn).await?;
+
+            assert_eq!(ok.unwrap(), "Success");
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn retry_on_io_error() -> anyhow::Result<()> {
+            let values = Mutex::new(vec![
+                Ok(TxId::from_str(
+                    "0x98f01c73c2062b55bba70966917a0839995e86abfadfff24534262d1c8b7a64e",
+                )),
```

Comment:
maybe we can extract this to a `tx_id` variable and later just use `assert_eq!(ok, tx_id);`
```suggestion
            let tx_id = TxId::from_str(
                "0x98f01c73c2062b55bba70966917a0839995e86abfadfff24534262d1c8b7a64e",
            );
            let values = Mutex::new(vec![
                Ok(tx_id),
```